connector/profilingmetricsconnector/frame.go

```diff
@@ -120,70 +195,171 @@ func fetchFrameInfo(dictionary pprofile.ProfilesDictionary,
 		}
 	}
 	if frameType == "" {
+		// Return error if we can't extract frame type in order to signal caller
+		// to stop further processing, as the current sample can not be relied upon.
 		return frameInfo{}, fmt.Errorf("fetchFrameInfo: empty frame type")
 	}
 
-	var fname string
-	if frameType == frameTypeNative {
+	fileName := ""
+	funcName := ""
+
+	// Errors related to extracting filename and function name are logged but not
+	// returned, as classification can still take place and the caller may continue
+	// to iterate frames.
+	switch frameType {
+	case frameTypeNative:
 		// Extract mapping filename
-		mappingTable := dictionary.MappingTable()
-		mtLen := mappingTable.Len()
-		mi := leaf.MappingIndex()
-		if int(mi) >= mtLen {
-			return frameInfo{}, fmt.Errorf("fetchFrameInfo: mi %d >= %d",
-				mi, mtLen)
-		}
-		mapFilenameStri := int(mappingTable.At(int(mi)).FilenameStrindex())
-		if mapFilenameStri >= strLen {
-			return frameInfo{}, fmt.Errorf("fetchFrameInfo: mapping.FilenameStridex %d >= %d",
-				mapFilenameStri, strLen)
-		}
-		fname = strTable.At(mapFilenameStri)
+		mapTable := dictionary.MappingTable()
+		mapTbLen := mapTable.Len()
+		mi := int(loc.MappingIndex())
+		if mi >= mapTbLen {
+			c.logger.Error("fetchFrameInfo",
+				zap.Int("mapIdx", mi),
+				zap.Int("mapTbLen", mapTbLen))
+			break
+		}
+		mapFileStrIdx := int(mapTable.At(int(mi)).FilenameStrindex())
+		if mapFileStrIdx >= strTbLen {
+			c.logger.Error("fetchFrameInfo",
+				zap.Int("mapFileStrIdx", mapFileStrIdx),
+				zap.Int("strTbLen", strTbLen))
+			break
+		}
+		fileName = strTable.At(mapFileStrIdx)
+	case frameTypeKernel:
+		// Extract function name
+		for _, ln := range loc.Line().All() {
+			funcIdx := int(ln.FunctionIndex())
+			if funcIdx >= funcTbLen {
+				c.logger.Error("fetchFrameInfo",
+					zap.Int("funcIdx", funcIdx),
+					zap.Int("funcTbLen", funcTbLen))
+				continue
+			}
+			fn := funcTable.At(funcIdx)
+			nameStrIdx := int(fn.NameStrindex())
+			if nameStrIdx >= strTbLen {
+				c.logger.Error("fetchFrameInfo",
+					zap.Int("nameStrIdx", nameStrIdx),
+					zap.Int("strTbLen", strTbLen))
+				continue
+			}
+			funcName = strTable.At(int(fn.NameStrindex()))
+			if funcName != "" {
+				break
+			}
+		}
 	}
 
 	// End validation
-	return frameInfo{typ: frameType, filename: fname}, nil
+	return frameInfo{
+		typ:      frameType,
+		fileName: fileName,
+		funcName: funcName,
+	}, nil
+}
+
+// classifyUser classifies a non-kernel frame by increasing relevant metric counts.
+// Must NOT be called for kernel frames.
+func classifyUser(fi frameInfo,
+	counts map[metric]int64,
+	nativeCounts map[attrInfo]int64,
+	multiplier int64) {
+	ft := fi.typ
+
+	// We have a distinct metricUser (even if there's no specific frame type for it)
+	// so that we can calculate the total number of stacktraces for any given time
+	// interval: metricUser.count + metricKernel.count
+	counts[metricUser] += multiplier
+	metric := allowedFrameTypes[ft]
```

Comment:
Maybe we should use a helper function on allowedFrameTypes that returns something for frame types that are not listed in allowedFrameTypes.

Reply:
This would be complicating the code more I think, isn't it much simpler to just validate once by checking if value exists in the array? 

We don't use any other value so the only reason we check the array is to determine if we should error / abort processing.